### PR TITLE
<one less ABT>

### PR DESCRIPTION
**Pull request by mitchellwrosen**
## Overview

This PR eliminates the second copy of the `ABT` type (and associated types like `ABT.Term`, `ABT.Var`), and deletes the conversions between "v1" ABT <-> "v2" ABT (now there's just one ABT defined in `U.Core.ABT`). Also, a few equivalent helper functions were deleted along the way.

Now `Unison.ABT` re-exports all of `U.Core.ABT`. It's not totally clear to me what the difference is (or should be) between these modules, but now at least there's less duplication ;)

### CODE REVIEW

File: codebase2/core/U/Core/ABT.hs [hunk 1]
```diff
@@ -104,7 +90,7 @@ visit f t = flip fromMaybe (f t) $ case out t of
 
 -- | Apply an effectful function to an ABT tree top down, sequencing the results.
 visit' ::
-  (Traversable f, Applicative g, Monad g, Ord v) =>
+  (Traversable f, Monad g, Ord v) =>
```
Comment (aryairani) [inline codebase2/core/U/Core/ABT.hs:144]: weird

**Review by pchiusano — APPROVED** [review]
This might speed up some operations too!

File: codebase2/core/U/Core/ABT.hs [hunk 2]
```diff
@@ -31,16 +22,62 @@ data ABT f v r
...
+
+instance (forall a. Show a => Show (f a), Show v) => Show (Term f v a) where
+  -- annotations not shown
+  showsPrec p (Term _ _ out) = case out of
+    Var v -> showParen (p >= 9) $ \x -> "Var " ++ show v ++ x
+    Cycle body -> ("Cycle " ++) . showsPrec p body
+    Abs v body -> showParen True $ (show v ++) . showString ". " . showsPrec p body
+    Tm f -> showsPrec p f
 
 amap :: (Functor f, Foldable f) => (a -> a') -> Term f v a -> Term f v a'
```
Comment (ChrisPenner) [inline codebase2/core/U/Core/ABT.hs:79]: This was here before, but does this really require `Foldable` to do an fmap?

File: codebase2/core/U/Core/ABT.hs [hunk 3]
```diff
@@ -162,10 +199,82 @@ pattern Cycle' vs t <- Term _ _ (Cycle (AbsN' vs t))
 pattern AbsN' :: [v] -> Term f v a -> Term f v a
 pattern AbsN' vs body <- (unabs -> (vs, body))
 
+{-# COMPLETE AbsN' #-}
```
Comment (ChrisPenner) [inline codebase2/core/U/Core/ABT.hs:202]: If this is COMPLETE, should it be a two-way pattern instead (using `abs` in the other direction)?

File: unison-core/src/Unison/ABT.hs [hunk 4]
```diff
@@ -203,9 +200,12 @@ amap' f t@(Term _ a out) = case out of
   Cycle r -> cycle' (f t a) (amap' f r)
   Abs v body -> abs' (f t a) v (amap' f body)
 
--- | Modifies the annotations in this tree
-instance Functor f => Functor (Term f v) where
-  fmap f (Term fvs a sub) = Term fvs (f a) (fmap (fmap f) sub)
+-- amap :: (Functor f, Foldable f) => (a -> a') -> Term f v a -> Term f v a'
```
Comment (ChrisPenner) [inline unison-core/src/Unison/ABT.hs:203]: Is this Leftover maybe? 😄

**Review by ChrisPenner — APPROVED** [review]
AWESOME.

Love to see this sort of duplication getting squashed, we've got a fair amount of it 😄

File: codebase2/core/U/Core/ABT.hs (hunk 2 above)
Comment (mitchellwrosen) [reply codebase2/core/U/Core/ABT.hs:79]: Nope!

File: unison-core/src/Unison/ABT.hs (hunk 4 above)
Comment (mitchellwrosen) [reply unison-core/src/Unison/ABT.hs:203]: Whoops, thanks

File: codebase2/core/U/Core/ABT.hs (hunk 3 above)
Comment (mitchellwrosen) [reply codebase2/core/U/Core/ABT.hs:202]: It could be bidirectional, yeah; I made it `COMPLETE` to satisfy GHC, I think

**Comment by mitchellwrosen** [comment]
Oops I merged this before pushing those fixes

**Pull request merged**